### PR TITLE
cmd/bosun: png/svg graph style changes

### PR DESCRIPTION
**Pull request by kylebrandt**
Before:

warning__os_cpu__on_skippygohungles_-_kyle_stackoverflow_com_-_stack_exchange_mail

After:

warning__os_cpu__on_skippygohungles_-_kyle_stackoverflow_com_-_stack_exchange_mail

The new argument to Graph is a list of units for the axis. It is vargs so this isn't breaking change (you can omit the second argument). The "title" has been moved to notes at the bottom of the graph, and moved outside of the svg/png. This means they can be copied and pasted when it comes in an email, and there is no issue with text being cut off.

### CODE REVIEW

**Comment by captncraig** [comment]
Do we need to update docs for this?

File: cmd/bosun/sched/template.go [hunk 1]
```diff
@@ -276,16 +276,19 @@ func (c *Context) EvalAll(v interface{}) (interface{}, error) {
...
+	res, exprText, err := c.eval(v, filter, true, 1000)
 	if err != nil {
 		return nil, err
 	}
 	var buf bytes.Buffer
 	const width = 800
 	const height = 600
+	footerHTML := fmt.Sprintf(`<small>Query: %s<br>Time: %s</small>`,
+		template.HTMLEscapeString(fmt.Sprint(exprText)),
+		fmt.Sprint(c.runHistory.Start.Format(time.RFC3339)))
```
Comment (captncraig) [inline cmd/bosun/sched/template.go]: Why are we `Sprint`ing a string inside of a `Sprintf`?

File: cmd/bosun/sched/template.go [hunk 2]
```diff
@@ -295,26 +298,30 @@ func (c *Context) graph(v interface{}, filter bool) (interface{}, error) {
...
 	}
+	buf.WriteString(footerHTML)
 	return template.HTML(buf.String()), nil
 }
 
 // Graph returns an SVG for the given result (or expression, for which it gets the result)
 // with same tags as the context's tags.
-func (c *Context) Graph(v interface{}) (interface{}, error) {
-	return c.graph(v, true)
+func (c *Context) Graph(v interface{}, unit ...string) (interface{}, error) {
```
Comment (captncraig) [inline cmd/bosun/sched/template.go]: When would I pass multiple unit args? Is this just vararg to make it a non-breaking change? I would rather break people than use var args as optional parameters.

Comment (gbrayut) [inline cmd/bosun/sched/template.go]: If you break it you get to up date all the existing templates :-P

Another option is a new Graph2 or GraphMeta function with the new signature

File: cmd/bosun/sched/template.go (hunk 1 above)
Comment (kylebrandt) [inline cmd/bosun/sched/template.go]: cause stupid.

**Comment by kylebrandt** [comment]
@captncraig Yes should update docs, but I don't know how the new method works.

File: cmd/bosun/sched/template.go (hunk 2 above)
Comment (kylebrandt) [inline cmd/bosun/sched/template.go]: It is an optional argument, so it is nice to not have to specify an empty string if I don't have to.

**Comment by captncraig** [comment]
lgtm

File: cmd/bosun/sched/chart.go [hunk 3]
```diff
@@ -33,17 +31,20 @@ func Autostyle(i int) chart.Style {
...
-		LineStyle:  chart.SolidLine,
-		LineWidth:  1,
-		LineColor:  c,
+		Symbol:      0,
+		SymbolSize:  0.2,
+		SymbolColor: c,
+		LineStyle:   chart.SolidLine,
+		LineWidth:   1,
+		LineColor:   c,
+		Font:        chart.Font{Size: chart.NormalFontSize},
```
Comment (mjibson) [inline cmd/bosun/sched/chart.go]: Are you sure this isn't the default?

File: cmd/bosun/sched/chart.go [hunk 4]
```diff
@@ -33,17 +31,20 @@ func Autostyle(i int) chart.Style {
 	c := chartColors[i%len(chartColors)]
 	return chart.Style{
 		// 0 uses a default
-		SymbolSize: 0.00001,
-		LineStyle:  chart.SolidLine,
-		LineWidth:  1,
-		LineColor:  c,
+		Symbol:      0,
```
Comment (mjibson) [inline cmd/bosun/sched/chart.go]: This is the default. Could we remove it?

File: cmd/bosun/sched/template.go [hunk 5]
```diff
@@ -276,16 +276,19 @@ func (c *Context) EvalAll(v interface{}) (interface{}, error) {
...
+func (c *Context) graph(v interface{}, unit string, filter bool) (interface{}, error) {
+	res, exprText, err := c.eval(v, filter, true, 1000)
 	if err != nil {
 		return nil, err
 	}
 	var buf bytes.Buffer
 	const width = 800
 	const height = 600
+	footerHTML := fmt.Sprintf(`<small>Query: %s<br>Time: %s</small>`,
+		template.HTMLEscapeString(fmt.Sprint(exprText)),
```
Comment (mjibson) [inline cmd/bosun/sched/template.go]: `exprText` is already a string, no need to convert it

File: cmd/bosun/sched/template.go [hunk 6]
```diff
@@ -295,26 +298,28 @@ func (c *Context) graph(v interface{}, filter bool) (interface{}, error) {
...
 	}
+	buf.WriteString(footerHTML)
 	return template.HTML(buf.String()), nil
 }
 
 // Graph returns an SVG for the given result (or expression, for which it gets the result)
 // with same tags as the context's tags.
-func (c *Context) Graph(v interface{}) (interface{}, error) {
-	return c.graph(v, true)
+func (c *Context) Graph(v interface{}, unit string) (interface{}, error) {
```
Comment (mjibson) [inline cmd/bosun/sched/template.go]: Having `unit` as the last argument means you can't do some things you could before with the go templates (see the [pipelines section](http://golang.org/pkg/text/template/#hdr-Pipelines) in the godocs). For example, here's a line copied from our prod.conf:

`{{printf "q(\"avg:rate{counter,,1}:os.cpu{host=%s}\", \"1d\", \"\")" (.Group.host) | .Graph}}`

This won't work anymore because you can only pass the last argument of a function via `|`.

One partial fix is to make `Graph` and `GraphAll` take a `...interface{}`. Then we could check the types and counts of things and handle them correctly. This would not break any existing syntax, however you still couldn't pass a y_label string to the above call.

One other possibility is to swap the order of the arguments.

We could also combine both solutions: swap and allow varargs. If the number of arguments is 1, then it's an expression. If it's 2, the first is a unit, the second is the expression.

File: docs/configuration.md [hunk 7]
```diff
@@ -131,8 +131,8 @@ Templates are the message body for emails that are sent when an alert is trigger
 * Eval(string): executes the given expression and returns the first result with identical tags, or `nil` tags if none exists, otherwise `nil`.
 * EvalAll(string): executes the given expression and returns all results.
 * GetMeta(metric, name, tags): Returns metadata data for the given combination of metric, metadata name, and tag. `metric` and `name` are strings. `tags` may be a tag string (`"tagk=tagv,tag2=val2"`) or a tag set (`.Group`). If If `name` is the empty string, a slice of metadata matching the metric and tag is returned. Otherwise, only the metadata value is returned for the given name, or `nil` for no match.
-* Graph(string): returns an SVG graph of the expression with identical tags
-* GraphAll(string): returns an SVG graph of the expression
+* Graph(expression, y_label): returns an SVG graph of the expression with tags identical to the alert instance. `expression` and `y_label` are strings. The second argument was introduced in v0.2.1.
```
Comment (mjibson) [inline docs/configuration.md]: I'm opposed to putting version notices in until we have a specific strategy for them. AngularJS has a dropdown on the side where you can select exactly the version you are running. Since we are going to soon have numbered releases, we could do that. Where did you get `0.2.1` from?

Comment (kylebrandt) [inline docs/configuration.md]: @mjibson Craig said that this would be release under that version.

File: cmd/bosun/sched/template.go (hunk 6 above)
Comment (kylebrandt) [inline cmd/bosun/sched/template.go]: Is it important that we support pipes for that command? What is the benefit besides possibly better syntax? Also I think that example might just be left over from before Graph was context aware.

File: cmd/bosun/web/chart.go [hunk 8]
```diff
@@ -122,7 +122,7 @@ func Graph(t miniprofiler.Timer, w http.ResponseWriter, r *http.Request) (interf
 	}
 	if _, present := r.Form["png"]; present {
 		c := chart.ScatterChart{
-			Title: fmt.Sprintf("%v - %v", oreq.Start, oreq.End),
+			Title: fmt.Sprintf("%v - %v", oreq.Start, queries),
```
Comment (mjibson) [inline cmd/bosun/web/chart.go]: Why this change?

File: docs/configuration.md [hunk 9]
```diff
@@ -131,8 +131,8 @@ Templates are the message body for emails that are sent when an alert is trigger
 * Eval(string): executes the given expression and returns the first result with identical tags, or `nil` tags if none exists, otherwise `nil`.
 * EvalAll(string): executes the given expression and returns all results.
 * GetMeta(metric, name, tags): Returns metadata data for the given combination of metric, metadata name, and tag. `metric` and `name` are strings. `tags` may be a tag string (`"tagk=tagv,tag2=val2"`) or a tag set (`.Group`). If If `name` is the empty string, a slice of metadata matching the metric and tag is returned. Otherwise, only the metadata value is returned for the given name, or `nil` for no match.
-* Graph(string): returns an SVG graph of the expression with identical tags
-* GraphAll(string): returns an SVG graph of the expression
+* Graph(expression, y_label): returns an SVG graph of the expression with tags identical to the alert instance. `expression` is a string or an expression and `y_label` is a string. `y_label` is an optional argument.
```
Comment (mjibson) [inline docs/configuration.md]: Perhaps use `[, y_label]` to also indicate that it's optional?

File: cmd/bosun/web/chart.go (hunk 8 above)
Comment (kylebrandt) [inline cmd/bosun/web/chart.go]: End was often nil and it wasn't handled. I'd vote for yanking out this feature all together.
